---
title: "Microbiome data analysis with microViz"
subtitle: "Transformations, PCA, & Differential Abundance"
author: "David Barnett"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: paper
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: true
---

# Setup

```{r setup}
knitr::opts_chunk$set(echo = TRUE, fig.width = 6, fig.height = 4, dpi = 120)
knitr::opts_knit$set(root.dir = '/xdisk/bhurwitz/bh_class/bonnie/exercises/19_pca_taxon_stats') # Change your NETID
```

```{r message=FALSE}
library(seriation)
library(dplyr)
library(purrr)
library(ggplot2)
library(phyloseq)
library(microViz)
library(shiny)
```

```{r}
shao19 <- readRDS('/xdisk/bhurwitz/bh_class/data/microviz/shao19.rds')
shao4 <- shao19 %>% ps_filter(family_role == "child", infant_age == 4)
```

# Beta diversity (part 4)

## PCA

Principal Components Analysis.

Quite similar to Principal Co-ordinates Analysis.

In fact, PCA produces equivalent results to PCoA with euclidean distances.
So let's perform the CLR-transform first and check PCA and euclidean PCoA are the same.

```{r}
shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  tax_transform(rank = "genus", trans = "clr", zero_replace = "halfmin") %>%
  dist_calc(dist = "euclidean") %>%
  ord_calc(method = "PCoA") %>%
  ord_plot(alpha = 0.6, size = 2, color = "birth_mode") +
  theme_classic(12) +
  coord_fixed(0.7) +
  scale_y_reverse()
```

```{r}
shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  tax_transform(rank = "genus", trans = "clr", zero_replace = "halfmin") %>%
  ord_calc(method = "PCA") %>%
  ord_plot(alpha = 0.6, size = 2, color = "birth_mode") +
  theme_classic(12) +
  coord_fixed(0.7)
```

So why is PCA interesting for us?
Because the Principal components are built directly from a (linear) combination of the original features.

That means we know how much each taxon contributes to each PC axis, and we can plot this information (loadings) as arrows, alongside the sample points.

```{r}
pca <- shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  tax_transform(rank = "genus", trans = "clr", zero_replace = "halfmin") %>%
  ord_calc(method = "PCA") %>%
  ord_plot(
    alpha = 0.6, size = 2, color = "birth_mode",
    plot_taxa = 1:6, tax_vec_length = 0.275,
    tax_lab_style = tax_lab_style(
      type = "text", max_angle = 90, aspect_ratio = 0.7,
      size = 3, fontface = "bold"
    ),
  ) +
  theme_classic(12) +
  coord_fixed(0.7, clip = "off")
pca
```

How to interpret the taxa loading vectors?
Cautiously.

There are caveats and nuance to the interpretation of these plots, which are called PCA bi-plots, and you can read more about here: <https://sites.google.com/site/mb3gustame/indirect-gradient-analysis/pca>

The relative length and direction of an arrow indicates how much that taxon contributes to the variation on each visible PC axis, e.g.
Variation in Enterococcus contributes quite a lot to variation along the PC2 axis.

This allows you to infer that samples positioned at the bottom of the plot will tend to have higher relative abundance of Enterococcus than samples at the top of the plot.

Interestingly, samples on the right of the plot (which tend to be vaginally-delivered infants) seem to have relatively more Bifidobacterium, Bacteroides and Escherichia, whilst the C-section born infants have relatively more Veillonella.


*(Side note, Phocaeicola were considered part of Bacteroides until this year!)*

You might have already noticed this pattern, when exploring and making barplots interactively with `ord_explore` earlier.

### Iris plot

We can make another kind of barplot now, using the PCA information to order our samples in a circular layout.

```{r}
iris <- shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  tax_transform(rank = "genus", trans = "clr", zero_replace = "halfmin") %>%
  ord_calc(method = "PCA") %>%
  ord_plot_iris(
    tax_level = "genus", n_taxa = 12, other = "Other",
    anno_colour = "birth_mode", 
    anno_colour_style = list(alpha = 0.6, size = 0.6, show.legend = FALSE)
  )
iris
```

```{r, fig.height=5, fig.width=10}
patchwork::wrap_plots(pca, iris, nrow = 1, guides = "collect")
```

# Taxon stats

From the PCA loadings and barplots above, we have some strong suspicions about which taxa have a higher relative abundance in vaginally delivered infants than in c-section delivered infants, and vice versa, but we can also statistically test this.
This is often called "differential abundance" testing, in the style of "differential expression" testing from the transcriptomics field.

```{r}
shao4 %>%
  comp_barplot(
    tax_level = "genus", n_taxa = 12, facet_by = "birth_mode",
    label = NULL, bar_outline_colour = NA
  ) +
  coord_flip() +
  theme(axis.ticks.y = element_blank())
```

## Model one taxon

We will start by creating a linear regression model for one genus, Bacteroides.
We will transform the count data by first making it proportions, and then taking the binary logarithm, log2, after adding a pseudocount.

```{r}
bacteroidesRegression1 <- shao4 %>%
  tax_transform("compositional", rank = "genus") %>%
  tax_model(
    type = "lm", rank = "genus",
    trans = "log2", trans_args = list(zero_replace = "halfmin"),
    taxa = "Bacteroides", variables = "birth_mode",
    return_psx = FALSE
  ) %>%
  pluck(1)

# looking at the regression results
summary(bacteroidesRegression1)
confint(bacteroidesRegression1)
broom::tidy(bacteroidesRegression1, conf.int = TRUE)
```

<details>

<summary>

**Click here for optional `ggplot2` extension exercise:**

</summary>

Starting from a dataframe like the one produced by the code below, plot:

1.  Easy: The percentage prevalence of Bacteroides in each birth_mode group
2.  Medium: The distribution of relative abundance of Bacteroides in each birth_mode group, omitting zeros, on a log2 scale
3.  Hard: Do task 1 or 2 for for several taxa in one plot - (hint: `pivot_longer`)

``` r
shao4 %>%
  tax_transform("compositional", rank = "genus") %>%
  ps_get() %>%
  ps_otu2samdat(taxa = "Bacteroides") %>%
  samdat_tbl()
```

</details>

We can fit a model with covariates, as we did for PERMANOVA.
We are going to convert the categorical variables into indicator (dummy) variables, and scale the continuous covariates to 0 mean and SD 1 (z-scores).
You'll see this will make our subsequent plots easier to interpret later.

```{r}
shao4 <- shao4 %>%
  ps_mutate(
    C_section = if_else(birth_mode == "c_section", true = 1, false = 0),
    Female = if_else(sex == "female", true = 1, false = 0),
    Birth_weight_Z = scale(birth_weight, center = TRUE, scale = TRUE),
    Reads_Z = scale(number_reads, center = TRUE, scale = TRUE)
  )
```

```{r}
bacteroidesRegression2 <- shao4 %>%
  tax_transform("compositional", rank = "genus") %>%
  tax_model(
    type = "lm", rank = "genus",
    trans = "log2", trans_args = list(zero_replace = "halfmin"),
    taxa = "Bacteroides", variables = c("C_section", "Female", "Birth_weight_Z", "Reads_Z"),
    return_psx = FALSE
  ) %>%
  pluck(1)

# looking at the regression results
summary(bacteroidesRegression1)
confint(bacteroidesRegression1)
broom::tidy(bacteroidesRegression1, conf.int = TRUE)
```

### Many methods

This method is what MaAsLin2 does by default (except they call the compositional transformation "Total Sum Scaling (TSS)"). This is quite a straightforward method, so we will stick with this for today, but many statistical methods have been developed for differential abundance analyses.

Microbiome abundance data are quite awkward, statistically speaking, due to their sparseness and compositionality.
Each successive method claims to handle some aspect of this awkwardness "better" than previous methods.

The aim is to have a method with adequate power to detect true associations, whilst controlling the type 1 error rate, the "false positive" reporting of associations that are not "truly" present.

Results are surprisingly inconsistent across the different methods, as demonstrated this year in a [fascinating analysis by Jacob Nearing and colleagues](https://www.nature.com/articles/s41467-022-28034-z).

**What to do?**

-   Filter out the noise & interpret results with caution! use multiple testing corrections
-   Remember it's all relative (abundance)
-   Try multiple methods and/or use same method as previous study if replicating 
    - Avoid Lefse and edgeR? 
    - Beware: Not all methods allow covariate adjustment & few allow random effects (for time-series)

## Model all the taxa!

We're not normally interested in just one taxon!
And often it's also hard to decide which taxonomic rank we are most interested in modelling!


*Lower ranks like species or ASVs give better resolution but also more sparsity and classification uncertainty...* *Higher ranks e.g. classes, could also be more powerful if you think most taxa within that class will follow a similar pattern.*


So now we will fit a similar model for almost\* every taxon at every rank we have available, from phylum down to species.

![](./images/model-all-the-taxa.jpg){width="50%"}


\*We'll actually filter out species with a prevalence of less than 10%.

```{r}
# The code for `taxatree_models` is quite similar to tax_model. 
# However, you might need to run `tax_prepend_ranks` to ensure that each taxon at each rank is always unique. 
shaoModels <- shao4 %>%
  tax_prepend_ranks() %>%
  tax_transform("compositional", rank = "species", keep_counts = TRUE) %>%
  tax_filter(min_prevalence = 0.1, undetected = 0, use_counts = TRUE) %>%
  taxatree_models(
    type = lm, trans = "log2", trans_args = list(zero_replace = "halfmin"),
    ranks = c("phylum", "class", "order", "family", "genus", "species"),
    variables = c("C_section", "Female", "Birth_weight_Z", "Reads_Z")
  )
shaoModels
```

*Why filter the taxa? It's less likely that we are interested in rare taxa, and models of rare taxon abundances are more 
likely to be unreliable. Reducing the the number of taxa modelled also makes the process faster and 
makes visualizing the results easier!*

### Getting stats from the models

Next we will get a data.frame containing the regression coefficient estimates, test statistics and corresponding p values from all these regression models.

```{r}
shaoStats <- taxatree_models2stats(shaoModels)
shaoStats
shaoStats %>% taxatree_stats_get()
```

### Adjusting p values

As we have performed a lot of statistical tests here, it is quite possible that could we find some significant p-values by chance alone.

So we should correct for multiple testing / control the false discovery rate or family-wise error rate.

Instead of applying these adjustment methods across all 86 taxa models at all ranks, the default behaviour is to control the family-wise error rate per taxonomic rank.

```{r}
shaoStats <- shaoStats %>% taxatree_stats_p_adjust(method = "BH", grouping = "rank")
# notice the new variable
shaoStats %>% taxatree_stats_get()
```

## Plot all the taxatree_stats!

`taxatree_plots()` allows you to plot statistics (e.g. point estimates and significance) from all of the taxa models onto a tree layout.
The taxon models are organised by rank, radiating out from the central root node from e.g.
Phyla around the center to Species in the outermost ring.

`taxatree_plots()` itself returns a list of plots, which you can arrange into one figure with the [`patchwork`](https://patchwork.data-imaginist.com/) package for example (and/or [`cowplot`](https://wilkelab.org/cowplot/articles/plot_grid.html)).

```{r, fig.width=6, fig.height=5}
shaoStats %>% 
  taxatree_plots(node_size_range = c(1, 3), sig_stat = 'p.adj.BH.rank') %>% 
  patchwork::wrap_plots(ncol = 2, guides = "collect")
```

### Taxatree Key

But how do we know which taxa are which nodes?
We can create a labelled grey tree with `taxatree_plotkey()`.
This labels only some of the taxa based on certain conditions that we specify.

```{r fig.height=4, fig.width=4.5, warning=FALSE}
set.seed(123) # label position 
key <- shaoStats %>% 
  taxatree_plotkey(
    taxon_renamer = function(x) stringr::str_remove(x, "[pfg]: "),
    # conditions below, for filtering taxa to be labelled
    rank == "phylum" | rank == "genus" & prevalence > 0.2
    # all phyla are labelled, and all genera with a prevalence of over 0.2
  ) 
key
```

You can do more with these trees to customise them to your liking.
See an extended tutorial [here on the microViz website](https://david-barnett.github.io/microViz/articles/web-only/modelling-taxa.html#plot-all-the-taxatree_stats): including how to directly label taxa on the colored plots, change the layout and style of the trees, and even how to use a different regression modelling approach.

```{r}
# try it out!
```

# Session info

Records your package versions etc.
Useful for debugging / reproducing analysis.

```{r}
devtools::session_info()
```
