<!DOCTYPE html>
<html>
<head>
<title>06_01-assessing-read-quality.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h2 id="quality-control">Quality control</h2>
<p>We will now assess the quality of the sequence reads contained in our FASTQ files.</p>
 <a href="../fig/03-02-02.png">
  <img src="../fig/03-02-02.png" alt="Flow diagram that shows the steps: Sequence reads and Quality control." />
</a>
<h3 id="details-on-the-fastq-format">Details on the FASTQ format</h3>
<p>Although it looks complicated (and it is), we can understand the <a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ</a> format with a little decoding. Some rules about the format include the following:</p>
<table>
<thead>
<tr>
<th>Line</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>Always begins with '@' followed by the information about the read</td>
</tr>
<tr>
<td>2</td>
<td>The actual DNA sequence</td>
</tr>
<tr>
<td>3</td>
<td>Always begins with a '+' and sometimes contains the same info as in line 1</td>
</tr>
<tr>
<td>4</td>
<td>Has a string of characters which represent the quality scores; must have same number of characters as line 2</td>
</tr>
</tbody>
</table>
<p>We can view the first complete read in one of the files from our dataset using <code>head</code> to look at
the first four lines. But we have to decompress one of the files first using <code>gunzip</code>.</p>
<pre class="hljs"><code><div>$ cd /xdisk/bhurwitz/bh_class/YOU_NETID/exercises/data/untrimmed_fastq
$ gunzip JP4D_R1.fastq.gz
$ head -n 4 JP4D_R1.fastq

</div></code></pre>
<pre class="hljs"><code><div>@MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:12622:2006 1:N:0:CTCAGA
CCCGTTCCTCGGGCGTGCAGTCGGGCTTGCGGTCTGCCATGTCGTGTTCGGCGTCGGTGGTGCCGATCAGGGTGAAATCCGTCTCGTAGGGGATCGCGAAGATGATCCGCCCGTCCGTGCCCTGAAAGAAATAGCACTTGTCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCTCAGAATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAAGCAAACCTCTCACTCCCTCTACTCTACTCCCTT                                        
+                                                                                                
A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### 
</div></code></pre>
<p>Line 4 shows the quality of each nucleotide in the read. Quality is interpreted as the
probability of an incorrect base call (e.g., 1 in 10) or, equivalently, the base call
accuracy (e.g., 90%). Each nucleotide's numerical score's value is converted into a character code where every single character
represents a quality score for an individual nucleotide. This conversion allows the alignment of each individual nucleotide with its quality
score. For example, in the line
above, the quality score line is:</p>
<pre class="hljs"><code><div>A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### 
</div></code></pre>
<p>The numerical value assigned to each character depends on the
sequencing platform that generated the reads. The sequencing machine used to generate our data
uses the standard Sanger quality PHRED score encoding, using Illumina version 1.8 onwards.
Each character is assigned a quality score between 0 and 41, as shown in
the chart below.</p>
<pre class="hljs"><code><div>Quality encoding: !&quot;#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJ
                   |         |         |         |         |
Quality score:    01........11........21........31........41                                
</div></code></pre>
<p>Each quality score represents the probability that the corresponding nucleotide call is
incorrect. These probability values are the results of the base calling algorithm and depend on how
much signal was captured for the base incorporation. This quality score is logarithmically based, so a quality score of 10 reflects a
base call accuracy of 90%, but a quality score of 20 reflects a base call accuracy of 99%.
In this
<a href="https://drive5.com/usearch/manual/quality_score.html">link</a> you can find more information
about quality scores.</p>
<p>Looking back at our read:</p>
<pre class="hljs"><code><div>@MISEQ-LAB244-W7:156:000000000-A80CV:1:1101:12622:2006 1:N:0:CTCAGA
CCCGTTCCTCGGGCGTGCAGTCGGGCTTGCGGTCTGCCATGTCGTGTTCGGCGTCGGTGGTGCCGATCAGGGTGAAATCCGTCTCGTAGGGGATCGCGAAGATGATCCGCCCGTCCGTGCCCTGAAAGAAATAGCACTTGTCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCTCAGAATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAAGCAAACCTCTCACTCCCTCTACTCTACTCCCTT                                        
+                                                                                                
A&gt;&gt;1AFC&gt;DD111A0E0001BGEC0AEGCCGEGGFHGHHGHGHHGGHHHGGGGGGGGGGGGGHHGEGGGHHHHGHHGHHHGGHHHHGGGGGGGGGGGGGGGGHHHHHHHGGGGGGGGHGGHHHHHHHHGFHHFFGHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFFFFFFFFFFFFFFFFFFFFBFFFF@F@FFFFFFFFFFBBFF?@;@#################################### 
</div></code></pre>
<p>We can now see that there is a range of quality scores but that the end of the sequence is
very poor (<code>#</code> = a quality score of 2).</p>
<blockquote>
<h2 id="exercise-1-looking-at-specific-reads">Exercise 1: Looking at specific reads</h2>
<p>In the terminal, how would you show the ID and quality of the last read <code>JP4D_R1.fastq</code>?<br>
a) <code>tail JP4D_R1.fastq</code><br>
b) <code>head -n 4 JP4D_R1.fastq</code><br>
c) <code>more JP4D_R1.fastq</code><br>
d) <code>tail -n4 JP4D_R1.fastq</code><br>
e) <code>tail -n4 JP4D_R1.fastq | head -n2</code></p>
<p>Do you trust the sequence in this read?</p>
</blockquote>
<details>
  <summary markdown="span">Solution</summary>
  <ul>
<p>a) It shows the ID and quality of the last read but also unnecessary lines from previous reads.<br>
b) No. It shows the first read's info.<br>
c) It shows the text of the entire file.<br>
d) This option is the best answer as it only shows the last read's information.<br>
e) It does show the ID of the last read but not the quality.</p>
<pre class="hljs"><code><div>@MISEQ-LAB244-W7:156:000000000-A80CV:1:2114:17866:28868 1:N:0:CTCAGA

CCCGTTCTCCACCTCGGCGCGCGCCAGCTGCGGCTCGTCCTTCCACAGGAACTTCCACGTCGCCGTCAGCCGCGACACGTTCTCCCCCCTCGCATGCTCGTCCTGTCTCTCGTGCTTGGCCGACGCCTGCGCCTCGCACTGCGCCCGCTCGGTGTCGTTCATGTTGATCTTCACCGTGGCGTGCATGAAGCGGTTCCCGGCCTCGTCGCCACCCACGCCATCCGCGTCGGCCAGCCACTCTCACTGCTCGC

+

AA11AC1&gt;3@DC1F1111000A0/A///BB#############################################################################################################################################################################################################################          
</div></code></pre>
<p>This read has more consistent quality at its first than at the end
but still has a range of quality scores,
most of them are low. We will look at variations in position-based quality in just a moment.</p>
</details>
<p>In real life, you won't be assessing the quality of your reads by visually inspecting your FASTQ files. Instead, you'll use a software program to assess read quality and filter out poor reads. We'll first use a program called <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> to visualize the quality of our reads.
Later in our class, we'll use another program to filter out poor-quality reads called trimmomatic</p>
<h2 id="using-fastqc-to-look-at-quality">Using fastqc to look at quality</h2>
<p>First, let's see if you can use fastqc from a container on the HPC. Pull up a shell in the HPC on demand. Remember that to use any code, we need to start an interactive node. If we call FastQC, a long help page will be displayed on our screen.</p>
<pre class="hljs"><code><div>$ interactive  
$ apptainer run /contrib/singularity/shared/bhurwitz/fastqc-0.11.9.sif fastqc -h 
</div></code></pre>
<pre class="hljs"><code><div>            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

        fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam]
           [-c contaminant file] seqfile1 .. seqfileN

DESCRIPTION

    FastQC reads a set of sequence files and produces from each one a quality
    control report consisting of many different modules, each one of
    which will help to identify a different potential type of problem in your
    data.
.
.
.
</div></code></pre>
<h2 id="assessing-quality-using-fastqc">Assessing quality using FastQC</h2>
<p>FastQC has several features that can give you a quick impression of any problems your data may have, so you can consider these issues before moving forward with your analyses. Rather than looking at quality scores for each read, FastQC looks at quality collectively across all reads within a sample. The image below shows one FastQC-generated plot that indicates a very high-quality sample:</p>
 <a href="../fig/03-02-03.png">
  <img src="../fig/03-02-03.png" alt="Quality graph showing a boxplot for each sequence position in the x-axis, and the Phred score in the y-axis. The background is colored red for the Phred scores 0 to 20, yellow for the scores 20 to 28, and green for the scores 28 to 38. All of the boxes for each position are in the green area." />
</a>
<p>The x-axis displays the base position in the read, and the y-axis shows quality scores. In this
example, the sample contains reads that are 40 bp long. This length is much shorter than the reads we
are working on within our workflow. For each position, there is a box-and-whisker plot showing
the distribution of quality scores for all reads at that position. The horizontal red line
indicates the median quality score, and the yellow box shows the 1st to
3rd quartile range. This range means that 50% of reads have a quality score that falls within the
range of the yellow box at that position. The whiskers show the whole range covering
the lowest (0th quartile) to highest (4th quartile) values.</p>
<p>The quality values for each position in this sample do not drop much lower than 32, which is a high-quality score. The plot background is also color-coded to identify good (green),
acceptable (yellow) and bad (red) quality scores.</p>
<p>Now let's look at a quality plot on the other end of the spectrum.</p>
 <a href="../fig/03-02-04.png">
  <img src="../fig/03-02-04.png" alt="Graphic of boxplots, where the first ones are in the good range of scores of the y-axis and extend to the acceptable and bad ranges of scores toward the right of the x-axis" />
</a>
<p>The FastQC tool produces several other diagnostic plots to assess sample quality and the one plotted above. Here, we see positions within the read in which the boxes span a much more comprehensive range. Also, quality scores drop pretty low into the &quot;bad&quot; range, particularly on the tail end of the reads.</p>
<h2 id="running-fastqc">Running FastQC</h2>
<p>We will now assess the quality of the reads that we downloaded. First, make sure you're still in the <code>untrimmed_fastq</code> directory.</p>
<pre class="hljs"><code><div>$ cd /xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/data/untrimmed_fastq
</div></code></pre>
<blockquote>
<h2 id="exercise-2-looking-at-metadata-about-the-untrimmed-files">Exercise 2: Looking at metadata about the untrimmed-files</h2>
<p>To know which files have more data, you need to see metadata about untrimmed files. In files, metadata includes owners of the file, state of the write, read, and execute permissions, size, and modification date. Using the <code>ls</code> command, how would you get the size of the files in the <code>untrimmed_fastq\</code> directory?<br>
(Hint: Look at the options for the <code>ls</code> command to see how to show file sizes.)<br>
a) <code>ls -a</code><br>
b) <code>ls -S</code><br>
c) <code>ls -l</code><br>
d) <code>ls -lh</code><br>
e) <code>ls -ahlS</code></p>
</blockquote>
<details>
  <summary markdown="span">Solution</summary>
  <ul> 
<pre class="hljs"><code><div>a) No. The flag `-a` shows all the contents, including hidden files and directories, but not the sizes.  
b) No. The flag `-S` shows the content Sorted by size, starting with the most extensive file, but not the sizes.  
c) Yes. The flag `-l` shows the contents with metadata, including file size. Other metadata are permissions, owners, and modification dates.    
d) Yes. The flag `-lh` shows the content with metadata in a human-readable manner.  
e) Yes. The combination of all the flags shows all the contents with metadata, including hidden files, sorted by size.  
</div></code></pre>
<pre class="hljs"><code><div>$ ls -ahls
</div></code></pre>
<pre class="hljs"><code><div>-rw-r--r-- 1 user user  24M Nov 26 21:34 JC1A_R1.fastq.gz                      
-rw-r--r-- 1 user user  24M Nov 26 21:34 JC1A_R2.fastq.gz                      
-rw-r--r-- 1 user user 616M Nov 26 21:34 JP4D_R1.fastq              
-rw-r--r-- 1 user user 203M Nov 26 21:35 JP4D_R2.fastq.gz   
</div></code></pre>
<p>Four FASTQ files oscillate between 24M (24MB) to 616M. The largest file is JP4D_R1.fastq with 616M.</p>
</details>
<br>
<p>FastQC can accept multiple file names as input, and on both zipped and unzipped files,
so we can use the <code>\*.fastq*</code>wildcard to run FastQC on all FASTQ files in this directory.</p>
<pre class="hljs"><code><div>$ apptainer run /contrib/singularity/shared/bhurwitz/fastqc-0.11.9.sif fastqc *.fastq* 
</div></code></pre>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<pre class="hljs"><code><div>Started analysis of JC1A_R1.fastq.gz                                               
Approx 5% complete for JC1A_R1.fastq.gz                                            
Approx 10% complete for JC1A_R1.fastq.gz                                           
Approx 15% complete for JC1A_R1.fastq.gz                                           
Approx 20% complete for JC1A_R1.fastq.gz                                           
Approx 25% complete for JC1A_R1.fastq.gz                                           
Approx 30% complete for JC1A_R1.fastq.gz                                          
Approx 35% complete for JC1A_R1.fastq.gz  
</div></code></pre>
<p>It should take around five minutes for FastQC to run on all
four of our FASTQ files. When the analysis completes, your prompt
will return. So your screen will look something like this:</p>
<pre class="hljs"><code><div>Approx 80% complete for JP4D_R2.fastq.gz
Approx 85% complete for JP4D_R2.fastq.gz
Approx 90% complete for JP4D_R2.fastq.gz
Approx 95% complete for JP4D_R2.fastq.gz
Analysis complete for JP4D_R2.fastq.gz
$
</div></code></pre>
<p>The FastQC program has created several new files within our
<code>data/untrimmed_fastq/</code> directory.</p>
<pre class="hljs"><code><div>$ ls 
</div></code></pre>
<pre class="hljs"><code><div>JC1A_R1_fastqc.html  JC1A_R2_fastqc.html  JP4D_R1.fastq        JP4D_R2_fastqc.html  TruSeq3-PE.fa
JC1A_R1_fastqc.zip   JC1A_R2_fastqc.zip   JP4D_R1_fastqc.html  JP4D_R2_fastqc.zip
JC1A_R1.fastq.gz     JC1A_R2.fastq.gz     JP4D_R1_fastqc.zip   JP4D_R2.fastq.gz    
</div></code></pre>
<p>For each input FASTQ file, FastQC has created a <code>.zip</code> file and a
<code>.html</code> file. The <code>.zip</code> file extension indicates that this is
a compressed set of multiple output files. We'll be working
with these output files soon. The <code>.html</code> file is a stable webpage
displaying the summary report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these output files into our results directory.</p>
<pre class="hljs"><code><div>$ mv *.zip /xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/06_qc_trimming
$ mv *.html /xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/06_qc_trimming 
</div></code></pre>
<p>If we want to visualize these HTML files we can do so by copying them to our home directory and opening them via the Jupyter server.</p>
<p>First let's go to the results directory and copy the files into a directory in our home dir.</p>
<pre class="hljs"><code><div>$ cd /xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/06_qc_trimming
$ mkdir ~/exercise_06_01_fastqc
$ cp *.html ~/exercise_06_01_fastqc
</div></code></pre>
<h2 id="viewing-the-fastqc-results">Viewing the FastQC results</h2>
<p>By using the Jupyter server, we will be able to look at each of these HTML files by opening them in a web browser. Go to the HPC on demand site, and kick off a Jupyter notebook job. Once your instance is running, open it up and go to the exercise_06_01_fastqc folder.</p>
<p>Now we can open the 4 HTML files. You can double click on each html file to open it up in the web browser.</p>
<blockquote>
<h2 id="exercise-4-discuss-the-quality-of-sequencing-files">Exercise 4: Discuss the quality of sequencing files</h2>
<p>Discuss your results with a neighbor. Which sample(s) looks the best
per base sequence quality? Which sample(s) look the
worst?</p>
</blockquote>
<details>
  <summary markdown="span">Solution</summary>
  <ul> 
All of the reads contain usable data, but the quality decreases toward
the end of the reads. File JC1A_R2_fastqc shows the lowest quality.
</details>
<h2 id="decoding-the-other-fastqc-outputs">Decoding the other FastQC outputs</h2>
<p>We've now looked at quite a few &quot;Per base sequence quality&quot; FastQC graphs, but there are nine other graphs that we haven't talked about! Below we have provided a brief overview of interpretations for each plot. For more information, please see the FastQC documentation <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">here</a></p>
<ul>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/12%20Per%20Tile%20Sequence%20Quality.html"><strong>Per tile sequence quality</strong></a>: the machines that perform sequencing are divided into tiles. This plot displays patterns in base quality along these tiles. Consistently low scores are often found around the edges, but hot spots could also occur in the middle if an air bubble was introduced during the run.</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/3%20Per%20Sequence%20Quality%20Scores.html"><strong>Per sequence quality scores</strong></a>: a density plot of quality for all reads at all positions. This plot shows what quality scores are most common.</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html"><strong>Per base sequence content</strong></a>: plots the proportion of each base position over all of the reads. Typically, we expect to see each base roughly 25% of the time at each position, but this often fails at the beginning or end of the read due to quality or adapter content.</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/5%20Per%20Sequence%20GC%20Content.html"><strong>Per sequence GC content</strong></a>: a density plot of average GC content in each of the reads.</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/6%20Per%20Base%20N%20Content.html"><strong>Per base N content</strong></a>: the percent of times that 'N' occurs at a position in all reads. If there is an increase at a particular position, this might indicate that something went wrong during sequencing.</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/7%20Sequence%20Length%20Distribution.html"><strong>Sequence Length Distribution</strong></a>: the distribution of sequence lengths of all reads in the file. If the data is raw, there is often a sharp peak; however, if the reads have been trimmed, there may be a distribution of shorter lengths.</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/8%20Duplicate%20Sequences.html"><strong>Sequence Duplication Levels</strong></a>: a distribution of duplicated sequences. In sequencing, we expect most reads to only occur once. If some sequences are occurring more than once, it might indicate enrichment bias (e.g. from PCR). This might not be true if the samples are high coverage (or RNA-seq or amplicon).</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/9%20Overrepresented%20Sequences.html"><strong>Overrepresented sequences</strong></a>: a list of sequences that occur more frequently than would be expected by chance.</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/10%20Adapter%20Content.html"><strong>Adapter Content</strong></a>: a graph indicating where adapter sequences occur in the reads.</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/11%20Kmer%20Content.html"><strong>K-mer Content</strong></a>: a graph showing any sequences which may show a positional bias within the reads.</li>
</ul>
<h2 id="working-with-the-fastqc-text-output">Working with the FastQC text output</h2>
<p>Now that we've looked at our HTML reports and are getting a feel for the data, let's look more closely at the other output files. Go back to the shell.</p>
<pre class="hljs"><code><div>$ cd /xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/06_qc_trimming
$ ls 
</div></code></pre>
<pre class="hljs"><code><div>JC1A_R1_fastqc.html           JP4D_R1_fastqc.html                  
JC1A_R1_fastqc.zip            JP4D_R1_fastqc.zip                   
JC1A_R2_fastqc.html           JP4D_R2_fastqc.html                  
JC1A_R2_fastqc.zip            JP4D_R2_fastqc.zip 
</div></code></pre>
<p>Our <code>.zip</code> files are compressed files. Each contains multiple
different types of output files for a single input FASTQ file. To
view the contents of a <code>.zip</code> file, we can use the program <code>unzip</code>
to decompress these files. Let's try doing them all at once using a
wildcard.</p>
<pre class="hljs"><code><div>$ unzip *.zip 
</div></code></pre>
<pre class="hljs"><code><div>Archive:  JC1A_R1_fastqc.zip                                                       
caution: filename not matched:  JC1A_R2_fastqc.zip                                 
caution: filename not matched:  JP4D_R1_fastqc.zip                      
caution: filename not matched:  JP4D_R2_fastqc.zip  
</div></code></pre>
<p>This decompression didn't work. It identified the first file and got a warning message for the other <code>.zip</code> files. This is because <code>unzip</code>
expects to get only one zip file as input. We could go through and
unzip each file one at a time, but this is very time-consuming and
error-prone. Someday you may have 500 files to unzip!</p>
<p>A more efficient way is to use a <code>for</code> loop like we learned in the Command Line lesson to iterate through all of
our <code>.zip</code> files. Let's see what that looks like, and then we'll
discuss what we're doing with each line of our loop.</p>
<pre class="hljs"><code><div>$ for filename in *.zip
&gt; do
&gt; unzip $filename
&gt; done
</div></code></pre>
<p>In this example, the input is the four filenames (one filename for each of our <code>.zip</code> files).
Each time the loop iterates, it will assign a file name to the variable <code>filename</code>
and run the <code>unzip</code> command.
The first time through the loop,
<code>$filename</code> is <code>JC1A_R1_fastqc.zip</code>.
The interpreter runs the command <code>unzip</code> on <code>JC1A_R1_fastqc.zip</code>.
For the second iteration, <code>$filename</code> becomes
<code>JC1A_R2_fastqc.zip</code>. This time, the shell runs <code>unzip</code> on <code>JC1A_R2_fastqc.zip</code>.
It then repeats this process for the other <code>.zip</code> files in our directory.</p>
<p>When we run the <code>for</code> loop, you will see an output that starts like this:</p>
<pre class="hljs"><code><div>Archive:  JC1A_R1_fastqc.zip                                            
creating: JC1A_R1_fastqc/                                            
creating: JC1A_R1_fastqc/Icons/                                      
creating: JC1A_R1_fastqc/Images/                                    
inflating: JC1A_R1_fastqc/Icons/fastqc_icon.png                      
inflating: JC1A_R1_fastqc/Icons/warning.png                          
inflating: JC1A_R1_fastqc/Icons/error.png                            
inflating: JC1A_R1_fastqc/Icons/tick.png                             
inflating: JC1A_R1_fastqc/summary.txt                                
inflating: JC1A_R1_fastqc/Images/per_base_quality.png                
inflating: JC1A_R1_fastqc/Images/per_tile_quality.png                
inflating: JC1A_R1_fastqc/Images/per_sequence_quality.png            
inflating: JC1A_R1_fastqc/Images/per_base_sequence_content.png       
inflating: JC1A_R1_fastqc/Images/per_sequence_gc_content.png         
inflating: JC1A_R1_fastqc/Images/per_base_n_content.png              
inflating: JC1A_R1_fastqc/Images/sequence_length_distribution.png 
inflating: JC1A_R1_fastqc/Images/duplication_levels.png              
inflating: JC1A_R1_fastqc/Images/adapter_content.png                 
inflating: JC1A_R1_fastqc/fastqc_report.html                         
inflating: JC1A_R1_fastqc/fastqc_data.txt                            
inflating: JC1A_R1_fastqc/fastqc.fo  
</div></code></pre>
<p>The <code>unzip</code> program is decompressing the <code>.zip</code> files and creates
a new directory (with subdirectories) for each of our samples, to
store all of the different output that is produced by FastQC. There
are a lot of files here. We're going to focus on the
<code>summary.txt</code> file.</p>
<p>If you list the files in our directory, now you will see the following:</p>
<pre class="hljs"><code><div>$ ls 
</div></code></pre>
<pre class="hljs"><code><div>JC1A_R1_fastqc                  JP4D_R1_fastqc                                                     
JC1A_R1_fastqc.html             JP4D_R1_fastqc.html                                        
JC1A_R1_fastqc.zip              JP4D_R1_fastqc.zip                                                  
JC1A_R2_fastqc                  JP4D_R2_fastqc                                               
JC1A_R2_fastqc.html             JP4D_R2_fastqc.html                                             
JC1A_R2_fastqc.zip              JP4D_R2_fastqc.zip                                         
</div></code></pre>
<p>The <code>.html</code> files and the uncompressed <code>.zip</code> files are still present,
but now we also have a new directory for each sample. We can
see that it's a directory if we use the <code>-F</code> flag for <code>ls</code>.</p>
<pre class="hljs"><code><div>$ ls -F 
</div></code></pre>
<pre class="hljs"><code><div>JC1A_R1_fastqc/                  JP4D_R1_fastqc/                                                     
JC1A_R1_fastqc.html             JP4D_R1_fastqc.html                                        
JC1A_R1_fastqc.zip              JP4D_R1_fastqc.zip                                                  
JC1A_R2_fastqc/                  JP4D_R2_fastqc/                                               
JC1A_R2_fastqc.html             JP4D_R2_fastqc.html                                             
JC1A_R2_fastqc.zip              JP4D_R2_fastqc.zip                                         
</div></code></pre>
<p>Let's see what files are present within one of these output directories.</p>
<pre class="hljs"><code><div>$ ls -F JC1A_R1_fastqc/ 
</div></code></pre>
<pre class="hljs"><code><div>fastqc_data.txt  fastqc.fo  fastqc_report.html	Icons/	Images/  summary.txt
</div></code></pre>
<p>Use <code>less</code> to preview the <code>summary.txt</code> file for this sample.</p>
<pre class="hljs"><code><div>$ less JC1A_R1_fastqc/summary.txt 
</div></code></pre>
<pre class="hljs"><code><div>PASS    Basic Statistics        JC1A_R1.fastq.gz                     
FAIL    Per base sequence quality       JC1A_R1.fastq.gz             
PASS    Per tile sequence quality       JC1A_R1.fastq.gz             
PASS    Per sequence quality scores     JC1A_R1.fastq.gz             
WARN    Per base sequence content       JC1A_R1.fastq.gz             
FAIL    Per sequence GC content JC1A_R1.fastq.gz                     
PASS    Per base N content      JC1A_R1.fastq.gz                     
PASS    Sequence Length Distribution    JC1A_R1.fastq.gz             
FAIL    Sequence Duplication Levels     JC1A_R1.fastq.gz             
PASS    Overrepresented sequences       JC1A_R1.fastq.gz             
FAIL    Adapter Content JC1A_R1.fastq.gz  
</div></code></pre>
<p>The summary file gives us a list of tests that FastQC ran and tells
us whether this sample passed, failed, or is borderline (<code>WARN</code>). Remember, to quit from <code>less</code>, you must type <code>q</code>.</p>
<h2 id="documenting-our-work">Documenting our work</h2>
<p>We can make a record of the results we obtained for all our samples
by concatenating all of our <code>summary.txt</code> files into a single file
using the <code>cat</code> command. We'll call this <code>fastqc_summaries.txt</code> and store it to the same directory.</p>
<pre class="hljs"><code><div>$ cat */summary.txt &gt; fastqc_summaries.txt
</div></code></pre>
<blockquote>
<h2 id="exercise-4-quality-tests">Exercise 4: Quality tests</h2>
<p>Which samples failed at least one of FastQC's quality tests? What
test(s) did those samples failed</p>
</blockquote>
<details>
  <summary markdown="span">Solution</summary>
  <ul> 
<p>We can get the list of all failed tests using <code>grep</code>.</p>
<pre class="hljs"><code><div>$ cd /xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/06_qc_trimming
$ grep FAIL fastqc_summaries.txt
</div></code></pre>
<pre class="hljs"><code><div>FAIL    Per base sequence quality       JC1A_R1.fastq.gz             
FAIL    Per sequence GC content JC1A_R1.fastq.gz                     
FAIL    Sequence Duplication Levels     JC1A_R1.fastq.gz             
FAIL    Adapter Content JC1A_R1.fastq.gz                             
FAIL    Per base sequence quality       JC1A_R2.fastq.gz             
FAIL    Per sequence GC content JC1A_R2.fastq.gz                     
FAIL    Sequence Duplication Levels     JC1A_R2.fastq.gz             
FAIL    Adapter Content JC1A_R2.fastq.gz                             
FAIL    Per base sequence content       JP4D_R1.fastq     
FAIL    Adapter Content JP4D_R1.fastq                     
FAIL    Per base sequence quality       JP4D_R2.fastq.gz  
FAIL    Per base sequence content       JP4D_R2.fastq.gz  
FAIL    Adapter Content JP4D_R2.fastq.gz
</div></code></pre>
</details>
<blockquote>
<h2 id="quality-of-large-datasets">Quality of large datasets</h2>
<p>Explore <a href="https://multiqc.info/">MultiQC</a> if you want a tool that can show the quality of many samples at once.</p>
</blockquote>
<blockquote>
<h2 id="quality-encodings-vary">Quality Encodings Vary</h2>
<p>Although we've used a particular quality encoding system to demonstrate the interpretation of
read quality, different sequencing machines use different encoding systems. This means that
depending on which sequencer you use to generate your data, a <code>#</code> may not indicate
a poor quality base call.</p>
<p>This mainly relates to older Solexa/Illumina data.
However, it's essential that you know which sequencing platform was
used to generate your data to tell your quality control program which encoding
to use. If you choose the wrong encoding, you run the risk of throwing away good reads or
(even worse) not throwing away bad reads!</p>
</blockquote>

</body>
</html>
