{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5de7bac8",
   "metadata": {},
   "source": [
    "# Taxonomic Annotation of Reads\n",
    "\n",
    "This notebook will go through the workflow assigning taxonomy to reads from a microbiome\n",
    "\n",
    "Step 1: Assign taxonomy to reads (using 3 parts that run in a single job)\n",
    "1. Taxonomic assignment of reads using Kraken2\n",
    "2. Refinement of the taxonomic annotation using Bracken\n",
    "3. Separation of human and non-human reads using KrakenTools\n",
    "\n",
    "Step 2: Assign taxonomy to contigs from megahit (with 3 parts as above)\n",
    "\n",
    "Step 3: Assign taxonomy to contigs from metaspades (with 3 parts as above)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07ebed36",
   "metadata": {},
   "source": [
    "## Getting Started\n",
    "\n",
    "You will need to rerun this section each time you come back to this notebook to reset all directories and variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7a46521",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the variables for your netid and xfile\n",
    "netid = \"MY_NETID\"\n",
    "xfile = \"MY_XFILE\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d388ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Go into the working directory\n",
    "work_dir = \"/xdisk/bhurwitz/bh_class/\" + netid + \"/assignments/10_taxonomy\"\n",
    "%cd $work_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c830ee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the fastq directory. This is where we have our fastq files with human contam removed.\n",
    "fastq_dir = \"/xdisk/bhurwitz/bh_class/\" + netid + \"/assignments/07_contam_removal\"\n",
    "xfile_dir = \"/xdisk/bhurwitz/bh_class/\" + netid + \"/assignments/05_getting_data\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "380dffed",
   "metadata": {},
   "source": [
    "## Creating a config file\n",
    "Let's create a config file with all of the variables we will need in the scripts below. Then when we want to use these variables in the script, we will \"source\" the config file to set the variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "049cf341",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a config file with all of the variables you need\n",
    "# notice that now we are defining the tools here too, to make out scripts easier to read below\n",
    "# also, if we want to update to a newer version of the tool we can just edit here.\n",
    "!echo \"export NETID=$netid\" > config.sh\n",
    "!echo \"export XFILE=$xfile\" >> config.sh\n",
    "!echo \"export WORK_DIR=$work_dir\" >> config.sh\n",
    "!echo \"export XFILE_DIR=$xfile_dir\" >> config.sh\n",
    "!echo \"export FASTQ_DIR=$fastq_dir\" >> config.sh\n",
    "!echo \"export MEGAHIT_CONTIG_DIR=/xdisk/bhurwitz/bh_class/$netid/assignments/08_assembly/out_megahit\" >> config.sh\n",
    "!echo \"export KRAKEN2=/contrib/singularity/shared/bhurwitz/kraken2:2.1.3--pl5321hdcf5f25_0.sif\" >> config.sh\n",
    "!echo \"export BRACKEN=/contrib/singularity/shared/bhurwitz/bracken:2.8--py39h1f90b4d_1.sif\" >> config.sh\n",
    "!echo \"export KRAKENTOOLS=/contrib/singularity/shared/bhurwitz/krakentools:1.2--pyh5e36f6f_0.sif\" >> config.sh\n",
    "!echo \"export DB_DIR=/groups/bhurwitz/databases/kraken2/k2_pluspf_20230605\" >> config.sh\n",
    "!echo \"export KMER_SIZE=100\" >> config.sh\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "bfeaabfe",
   "metadata": {},
   "source": [
    "## Step 1: Assigning taxonomy to the reads\n",
    "\n",
    "In this step, we will assign taxonomy to each of the reads in our microbiome (where possible). This script involves multiple steps in the annotation process including:\n",
    "\n",
    "1. Running Kraken2 to assign reads to organisms by taxonomic rank. \n",
    "2. Running Bracken a tool that refines the Kraken2 output to try to reassign reads to higher species-level taxonomy\n",
    "3. Separating out the reads based on their taxonomy using KrakenTools. We do this at a high level human / non-human.\n",
    "\n",
    "The final part in this analysis looks for any remaining human contamination in the files, after we align to the human genome using bowtie2. Do you see any reads reported as human? Try \"Blasting\" these to see if they are indeed human reads that we missed by the alignment. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df7e3021",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a script to run Kraken, Bracken, and KrakenTools to assign taxonomy at the read-level.\n",
    "# A few important points:\n",
    "# 1. We are using the variables from the config file via the `source ./config.sh` command in the script.\n",
    "# 2. Kraken2 runs on each of the fastq files in the trimmed/human screened $FASTQ_DIR\n",
    "# 3. The results will be written into our $WORK_DIR\n",
    "# 4. Notice that we are asking for alot more resource (24 cores and 5G of memory per core), we are also asking for 10 hours of runtime.\n",
    "# But, runtime will likely be much less.\n",
    "my_code = '''#!/bin/bash\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --nodes=1             \n",
    "#SBATCH --time=10:00:00   \n",
    "#SBATCH --partition=standard\n",
    "#SBATCH --account=bh_class\n",
    "#SBATCH --array=0-7                         \n",
    "#SBATCH --output=10A_read_taxonomy-%a.out\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH --mem-per-cpu=5G  \n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "source $SLURM_SUBMIT_DIR/config.sh\n",
    "names=($(cat $XFILE_DIR/$XFILE))\n",
    "\n",
    "SAMPLE_ID=${names[${SLURM_ARRAY_TASK_ID}]}\n",
    "\n",
    "### reads with human removed to match to the reference database\n",
    "PAIR1=${FASTQ_DIR}/${SAMPLE_ID}_1.fastq.gz\n",
    "PAIR2=${FASTQ_DIR}/${SAMPLE_ID}_2.fastq.gz\n",
    "\n",
    "KRAKEN_OUTDIR=${WORK_DIR}/out_reads_taxonomy\n",
    "OUTDIR=${KRAKEN_OUTDIR}/${SAMPLE_ID}\n",
    "HUMAN_READ_DIR=${OUTDIR}/human_reads\n",
    "NONHUMAN_READ_DIR=${OUTDIR}/nonhuman_reads\n",
    "\n",
    "### create the outdir if it does not exist\n",
    "if [[ ! -d \"$KRAKEN_OUTDIR\" ]]; then\n",
    "  echo \"$KRAKEN_OUTDIR does not exist. Directory created\"\n",
    "  mkdir $KRAKEN_OUTDIR\n",
    "fi\n",
    "\n",
    "if [[ ! -d \"$OUTDIR\" ]]; then\n",
    "  echo \"$OUTDIR does not exist. Directory created\"\n",
    "  mkdir $OUTDIR\n",
    "fi\n",
    "\n",
    "if [[ ! -d \"$HUMAN_READ_DIR\" ]]; then\n",
    "  echo \"$HUMAN_READ_DIR does not exist. Directory created\"\n",
    "  mkdir $HUMAN_READ_DIR\n",
    "fi\n",
    "\n",
    "if [[ ! -d \"$NONHUMAN_READ_DIR\" ]]; then\n",
    "  echo \"$NONHUMAN_READ_DIR does not exist. Directory created\"\n",
    "  mkdir $NONHUMAN_READ_DIR\n",
    "fi\n",
    "\n",
    "# check input\n",
    "echo ${PAIR1}\n",
    "echo ${PAIR2}\n",
    "echo ${OUTDIR}\n",
    "\n",
    "apptainer run ${KRAKEN2} kraken2 --db ${DB_DIR} --paired \\\n",
    "  --classified-out ${OUTDIR}/cseqs#.fq --output ${OUTDIR}/kraken_results.txt \\\n",
    "  --report ${OUTDIR}/kraken_report.txt --use-names --threads ${SLURM_CPUS_PER_TASK} \\\n",
    "  ${PAIR1} ${PAIR2}\n",
    "\n",
    "# refine hits with Bracken\n",
    "REPORT=\"${OUTDIR}/kraken_report.txt\"\n",
    "RESULTS=\"${OUTDIR}/kraken_results.txt\"\n",
    "apptainer run ${BRACKEN} est_abundance.py -i ${REPORT} -o ${OUTDIR}/bracken_results.txt -k ${DB_DIR}/database${KMER_SIZE}mers.kmer_distrib\n",
    "\n",
    "# get human and non-human reads (microbial)\n",
    "TAXID=9606\n",
    "HUMAN_R1=\"${HUMAN_READ_DIR}/r1.fq\"\n",
    "HUMAN_R2=\"${HUMAN_READ_DIR}/r2.fq\"\n",
    "\n",
    "BRACKEN_REPORT=\"${OUTDIR}/kraken_report_bracken_species.txt\"\n",
    "BRACKEN_RESULTS=\"${OUTDIR}/bracken_results.txt\"\n",
    "\n",
    "apptainer run ${KRAKENTOOLS} extract_kraken_reads.py -k ${RESULTS} \\\n",
    " -r ${BRACKEN_REPORT} -s1 ${PAIR1} -s2 ${PAIR2} --taxid ${TAXID} \\\n",
    " -o ${HUMAN_R1} -o2 ${HUMAN_R2} --include-children --fastq-output \n",
    "\n",
    "gzip ${HUMAN_READ_DIR}/r1.fq\n",
    "gzip ${HUMAN_READ_DIR}/r2.fq\n",
    "\n",
    "### selects all reads NOT from a given set of Kraken taxids (and all children)\n",
    "\n",
    "NONHUMAN_R1=\"${NONHUMAN_READ_DIR}/r1.fq\"\n",
    "NONHUMAN_R2=\"${NONHUMAN_READ_DIR}/r2.fq\"\n",
    "\n",
    "apptainer run ${KRAKENTOOLS} extract_kraken_reads.py -k ${RESULTS} \\\n",
    " -r ${BRACKEN_REPORT} -s1 ${PAIR1} -s2 ${PAIR2} --taxid ${TAXID} \\\n",
    " -o ${NONHUMAN_R1} -o2 ${NONHUMAN_R2} --include-children \\\n",
    " --exclude --fastq-output \n",
    "\n",
    "gzip ${NONHUMAN_READ_DIR}/r1.fq\n",
    "gzip ${NONHUMAN_READ_DIR}/r2.fq\n",
    "\n",
    "echo \"Finished `date`\"\n",
    "\n",
    "'''\n",
    "\n",
    "with open('10A_read_taxonomy.sh', mode='w') as file:\n",
    "    file.write(my_code)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f8b30b51",
   "metadata": {},
   "source": [
    "Nice, you should now have all of the reads assigned to the organisms they are from."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9dfb0bf5",
   "metadata": {},
   "source": [
    "## Step 2: Assigning taxonomy to the contigs from Megahit\n",
    "\n",
    "In this step, we will assign taxonomy to each of the contigs from our megahit assembly (where possible). This script involves multiple steps in the annotation process (just as before for the reads) including:\n",
    "\n",
    "1. Running Kraken2 to assign contigs to organisms by taxonomic rank. \n",
    "2. Running Bracken a tool that refines the Kraken2 output to try to reassign contigs to higher species-level taxonomy\n",
    "3. Separating out the contigs based on their taxonomy using KrakenTools. We do this at a high level human / non-human."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67182c11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a script to run Kraken, Bracken, and KrakenTools to assign taxonomy at the contig-level for megahit.\n",
    "# A few important points:\n",
    "# 1. We are using the variables from the config file via the `source ./config.sh` command in the script.\n",
    "# 2. Kraken2 runs on the final.contigs.fa file in the contig directory\n",
    "# 3. The results will be written into our $WORK_DIR\n",
    "# 4. Notice that we are asking for alot more resource (24 cores and 5G of memory per core), we are also asking for 10 hours of runtime.\n",
    "# But, runtime will likely be much less.\n",
    "my_code = '''#!/bin/bash\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --nodes=1             \n",
    "#SBATCH --time=10:00:00   \n",
    "#SBATCH --partition=standard\n",
    "#SBATCH --account=bh_class\n",
    "#SBATCH --array=0-7                        \n",
    "#SBATCH --output=10B_contig_taxonomy-%a.out\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH --mem-per-cpu=5G  \n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "source $SLURM_SUBMIT_DIR/config.sh\n",
    "names=($(cat $XFILE_DIR/$XFILE))\n",
    "\n",
    "SAMPLE_ID=${names[${SLURM_ARRAY_TASK_ID}]}\n",
    "\n",
    "### contigs\n",
    "CONTIGS=${MEGAHIT_CONTIG_DIR}/${SAMPLE_ID}/final.contigs.fa\n",
    "\n",
    "KRAKEN_OUTDIR=${WORK_DIR}/out_megahit_taxonomy\n",
    "OUTDIR=${KRAKEN_OUTDIR}/${SAMPLE_ID}\n",
    "HUMAN_READ_DIR=${OUTDIR}/human_contigs\n",
    "NONHUMAN_READ_DIR=${OUTDIR}/nonhuman_contigs\n",
    "\n",
    "### create the outdir if it does not exist\n",
    "if [[ ! -d \"$KRAKEN_OUTDIR\" ]]; then\n",
    "  echo \"$KRAKEN_OUTDIR does not exist. Directory created\"\n",
    "  mkdir $KRAKEN_OUTDIR\n",
    "fi\n",
    "                                  \n",
    "if [[ ! -d \"$OUTDIR\" ]]; then\n",
    "  echo \"$OUTDIR does not exist. Directory created\"\n",
    "  mkdir $OUTDIR\n",
    "fi\n",
    "\n",
    "if [[ ! -d \"$HUMAN_READ_DIR\" ]]; then\n",
    "  echo \"$HUMAN_READ_DIR does not exist. Directory created\"\n",
    "  mkdir $HUMAN_READ_DIR\n",
    "fi\n",
    "\n",
    "if [[ ! -d \"$NONHUMAN_READ_DIR\" ]]; then\n",
    "  echo \"$NONHUMAN_READ_DIR does not exist. Directory created\"\n",
    "  mkdir $NONHUMAN_READ_DIR\n",
    "fi\n",
    "\n",
    "apptainer run ${KRAKEN2} kraken2 --db ${DB_DIR} --classified-out \\\n",
    " ${OUTDIR}/cseqs#.fa --output ${OUTDIR}/kraken_results.txt \\\n",
    " --report ${OUTDIR}/kraken_report.txt --use-names --threads ${SLURM_CPUS_PER_TASK} \\\n",
    " ${CONTIGS}\n",
    "\n",
    "# refine hits with Bracken\n",
    "REPORT=\"${OUTDIR}/kraken_report.txt\"\n",
    "RESULTS=\"${OUTDIR}/kraken_results.txt\"\n",
    "apptainer run ${BRACKEN} est_abundance.py -i ${REPORT} -o ${OUTDIR}/bracken_results.txt -k ${DB_DIR}/database${KMER_SIZE}mers.kmer_distrib\n",
    "\n",
    "# get human and non-human reads (microbial)\n",
    "TAXID=9606\n",
    "HUMAN_R1=\"${HUMAN_READ_DIR}/contigs.fa\"\n",
    "\n",
    "BRACKEN_REPORT=\"${OUTDIR}/kraken_report_bracken_species.txt\"\n",
    "BRACKEN_RESULTS=\"${OUTDIR}/bracken_results.txt\"\n",
    "\n",
    "apptainer run ${KRAKENTOOLS} extract_kraken_reads.py -k ${RESULTS} \\\n",
    " -r ${BRACKEN_REPORT} -s1 ${CONTIGS} --taxid ${TAXID} -o ${HUMAN_R1} \\\n",
    " --include-children \n",
    "\n",
    "if [[ -f \"${HUMAN_READ_DIR}/contigs.fa\" ]]; then\n",
    "    gzip ${HUMAN_READ_DIR}/contigs.fa\n",
    "fi\n",
    "\n",
    "### selects all reads NOT from a given set of Kraken taxids (and all children)\n",
    "\n",
    "NONHUMAN_R1=\"${NONHUMAN_READ_DIR}/contigs.fa\"\n",
    "\n",
    "apptainer run ${KRAKENTOOLS} extract_kraken_reads.py -k ${RESULTS} \\\n",
    " -r ${BRACKEN_REPORT} -s1 ${CONTIGS} --taxid ${TAXID} -o ${NONHUMAN_R1} \\\n",
    " --include-children --exclude \n",
    "\n",
    "if [[ -f \"${NONHUMAN_READ_DIR}/contigs.fa\" ]]; then\n",
    "    gzip ${NONHUMAN_READ_DIR}/contigs.fa\n",
    "fi\n",
    "\n",
    "echo \"Finished `date`\"\n",
    "\n",
    "'''\n",
    "\n",
    "with open('10B_contig_taxonomy.sh', mode='w') as file:\n",
    "    file.write(my_code)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "febfaf63",
   "metadata": {},
   "source": [
    "## Step 3: Putting it all together\n",
    "\n",
    "Once you have created the the run scripts, you are ready to put them together in a pipeline to run each of the steps one by one. Notice which steps are dependent on the others."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1751d9e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's create the launcher script to kick off our pipeline.\n",
    "\n",
    "my_code = '''#! /bin/bash\n",
    "\n",
    "# 10A_read_taxonomy: jid1 has no dependencies\n",
    "job1=$(sbatch 10A_read_taxonomy.sh)\n",
    "jid1=$(echo $job1 | sed 's/^Submitted batch job //')\n",
    "echo $jid1\n",
    "\n",
    "# 10B_contig_taxonomy: jid2 has no dependencies\n",
    "job2=$(sbatch 10B_contig_taxonomy.sh)\n",
    "jid2=$(echo $job2 | sed 's/^Submitted batch job //')\n",
    "echo $jid2\n",
    "\n",
    "'''\n",
    "\n",
    "with open('10_launch_pipeline.sh', mode='w') as file:\n",
    "    file.write(my_code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a83c89c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the pipeline script executable\n",
    "!chmod +x *.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e027e40a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now let's run it!\n",
    "!./10_launch_pipeline.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb693f16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can check if it is running using the squeue command\n",
    "# Check for all jobs under your netid\n",
    "# Note that this will take some time to run, so go get a coffee!\n",
    "!squeue --user=$netid"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83501ab5",
   "metadata": {},
   "source": [
    "## Final Step\n",
    "Copy your notebook to the current working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5b6190d",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp ~/be487-fall-2024/assignments/10_taxonomy/hw10_taxonomy.ipynb $work_dir"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
