---
title: "Microbiome data analysis with microViz"
subtitle: "Alpha Diversity"
author: "David Barnett"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: paper
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: true
editor_options: 
  markdown: 
    wrap: sentence
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 6, fig.height = 4, dpi = 120)
knitr::opts_knit$set(root.dir = '/xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/14_alpha_diversity') # change your NETID
```

```{bash, engine.opts='-l'}
# Let's make it so R can find the fonts we need
export FONTCONFIG_PATH=/etc/fonts
```

```{r}
dyn.load("/opt/ohpc/pub/apps/libpng/1.6.37/lib/libpng16.so.16")
```

```{r message=FALSE}
library(seriation)
library(dplyr)
library(purrr)
library(ggplot2)
library(phyloseq)
library(microViz)
library(shiny)
```

```{r}
mice <- readRDS('/xdisk/bhurwitz/bh_class/data/microviz/mice.rds')
mice <- tax_fix(mice, verbose = FALSE)
```

# Reminder about our example `mice` dataset

`mice` is 16S rRNA gene amplicon sequencing data, from the mouse antibiotics study This is a `phyloseq` S4 object, containing processed microbiota data from the mouse study.

<details>

<summary>**Click here for more details about the mice dataset:**</summary>

**The mice data**

The data originate from a study on the effects of oral antibiotic administration on flavivirus infection (<https://www.ncbi.nlm.nih.gov/pubmed/29590614>).
Sequence data was generated from extracted nucleic acid from stool samples collected from individually caged mice and amplified using primers specific for the V4 region using primers 515F/806R.

The study followed flavivirus infection after the following treatments:

1)  Koolaid: Antibiotics are provided to the mice via their drinking water. As many of the antibiotics taste bad, koolaid is added as a sweetener. Therefore, the appropriate control is water spiked and labelled *koolaid*.
2)  Ampicillin (Amp): <https://en.wikipedia.org/wiki/Ampicillin>
3)  Metronidazole (Met): <https://en.wikipedia.org/wiki/Metronidazole>
4)  Ampicillin + Metronidazole (Amp+Metro)

Treatments were supplied ad libitum for 2 weeks prior to viral infection and maintained for 2 weeks post-infection.
Primary outcome was mouse survival.
Each treatment group had two subgroups of mice that were either a) left uninfected as controls or b) infected with West Nile Virus via a subcutaneous foot pad injection.

</details>

# Alpha diversity

How diverse is the bacterial microbiome of each sample?

## Why is this interesting?

### Biologically

-   Lower gut microbiome diversity is related to worse health in adult humans.
-   Higher diversity ecosystems are often considered healthier, more mature, and more resilient to perturbation.
-   BUT: diverse == healthy does not hold for all ecosystems, e.g. early infant gut microbiome, so consider your own data and hypotheses carefully.

### Practically

Diversity indices provide a simple one number summary of each ecosystem: easy to compare and do stats with.

## Richness

The more the merrier.
The simplest measure is just counting, aka "Observed Richness".
Let's compute the observed richness and label each sample with it.

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>% 
  ps_calc_richness(rank = 'Genus', index = 'observed', varname = 'N genera') %>%
  comp_barplot(
    tax_level = "Genus", n_taxa = 12, label = 'N genera', bar_width = 0.7,
    sample_order = 'asis', merge_other = FALSE#, tax_transform_for_plot = 'identity'
  ) +
  coord_flip()
```

## Diversity

Richness and evenness matter.

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>% 
  ps_calc_diversity(rank = 'Genus', index = 'shannon') %>%
  ps_mutate(shannon_Genus = round(shannon_Genus, digits = 2)) %>% 
  comp_barplot(
    tax_level = "Genus", n_taxa = 12, label = 'shannon_Genus', bar_width = 0.7,
    sample_order = 'asis', merge_other = FALSE
  ) +
  coord_flip()
```

## Statistics with alpha diversity

So we have our alpha diversity values for this small subset of mice.

```{r, fig.width=3, fig.height=1}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>% 
  ps_calc_diversity(rank = 'Genus', index = 'shannon') %>%
  samdat_tbl() %>% 
  ggplot(aes(x = shannon_Genus, y = "Day 13\ncontrols")) + 
  geom_point(position = position_jitter(height = 0.2), alpha = 0.5) +
  labs(x = 'Shannon diversity (Genus)', y = NULL) +
  xlim(1, 2.5) +
  theme_bw() 
```

Let's calculate alpha diversity for all mice after antibiotic or control treatment, and make a comparison.
I suspect that the average gut microbiota diversity of the antibiotic exposed mice will differ from the control group's at day 3.

```{r}
# First compute a new variable aggregating all the control mice together
mice <- mice %>% 
  ps_mutate(antibiotics = treatment %in% c("Amp", "Metro", "AmpMetro"))
```

```{r, fig.width=4, fig.height=2.5}
mice %>% 
  ps_filter(treatment_days == 'D13') %>% 
  ps_calc_diversity(rank = 'Genus', index = 'shannon') %>%
  samdat_tbl() %>% 
  ggplot(aes(y = antibiotics, x = shannon_Genus)) +
  geom_boxplot(width = 0.3) + 
  geom_point(position = position_jitter(height = 0.2), alpha = 0.5) +
  theme_bw()
```

It looks like the antibiotics treated mice have lower gut microbiota diversity on average.
A simple statistical test supports this.

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13') %>% 
  ps_calc_diversity(rank = 'Genus', index = 'shannon') %>%
  samdat_tbl() %>% 
  wilcox.test(formula = shannon_Genus ~ antibiotics, data = .)
```

You can apply more complex statistical tests as you like, e.g. adjusting for covariates with linear regression, using `lm()`

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13') %>% 
  ps_calc_diversity(rank = 'Genus', index = 'shannon') %>%
  samdat_tbl() %>% 
  lm(formula = shannon_Genus ~ antibiotics + virus, data = .) %>% 
  summary()
```

Try it out for yourself at other time points?
Practice making plots and doing simple statistical tests.

What about richness?

# Exercises

<details>

<summary>**Click here for an exercise with an IBD dataset:**</summary>

**Inflammatory Bowel Disease study**

```{r}
ibd <- readRDS('/xdisk/bhurwitz/bh_class/data/microviz/ibd.rds') # not sure if we need to run this?
```

```{r}
ibd <- corncob::ibd_phylo %>% 
  tax_mutate(Species = NULL) %>% # ibd_phylo Species column was blank -> deleted
  ps_mutate(disease = ibd == 'ibd', ibd = NULL) # adds disease state indicator variable
```

`ibd` is another phyloseq object containing 16S rRNA gene amplicon sequencing data, from a 2012 study of Inflammatory Bowel Disease in children and young adults.

It is an "old" data: where they used 454 Pyrosequencing, and clustered the raw sequences into "OTUs".

Have a look at the data, like we did before for the `mice` dataset.

```{r}
ibd
```

```{r}
#
```

You can perform alpha diversity analysis: Try comparing the alpha diversity of the IBD patients against the healthy controls.

```{r}
#
```

</details>

<details>

<summary>**Click here for additional notes on richness and readcount**</summary>

**Additional notes on richness and readcount**

Simple approaches like Observed Richness are sensitive to what ecologists call "sampling effort".
For macroecologists, this is actually how much time/effort you spent trying to count all the organisms present in an ecosystem.
In our case, the amount of total reads obtained represents the sampling effort: more reads, more effort.
Indeed we can see that the samples with a much lower readcount have lower observed richness.

(Furthermore, as this richness estimate is based on a sample, and not the actual ecosystem, the richness estimate actually has quantifiable uncertainty too.)

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Amp') %>%
  ps_calc_richness(rank = 'Genus', index = 'observed', varname = 'N genera') %>%
  comp_barplot(
    tax_level = "Genus", n_taxa = 12, label = 'N genera', bar_width = 0.7,
    sample_order = 'asis', merge_other = FALSE, tax_transform_for_plot = 'identity'
  )
```

```{r, fig.width=5, fig.height=3}
mice %>% 
  ps_calc_richness(rank = 'Genus', index = 'observed', varname = 'genera') %>%
  ps_mutate(readcount = sample_sums(mice)) %>% 
  samdat_tbl() %>% 
  ggplot(aes(readcount, genera)) + 
  geom_point(alpha = 0.4, size = 2.5) +
  theme_bw(14)
```

**What to do:**

1)  **Simple solution**: Ignore the problem. Whilst you can't interpret the richness of any individual sample as being correct, it is still usually valid to compare richness across groups of samples, as the readcount variation is only random noise, and should be uncorrelated with your grouping variable (but do check this).
2)  **Harder solution**: Explore more rigorous methods like `breakaway` by Amy Willis and team. <https://www.frontiersin.org/articles/10.3389/fmicb.2019.02407/full>

</details>

# Session info

Records your package versions etc.
Useful for debugging / reproducing analysis.

```{r}
devtools::session_info()
```
