---
title: "Microbiome data analysis with microViz"
subtitle: "Dissimilarity, ordination, & filtering"
author: "David Barnett"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: paper
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: true
editor_options: 
  markdown: 
    wrap: sentence
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 6, fig.height = 4, dpi = 120)
knitr::opts_knit$set(root.dir = '/xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/13_microviz') # Change your NETID
```

```{bash, engine.opts='-l'}
# Let's make it so R can find the fonts we need
export FONTCONFIG_PATH=/etc/fonts
```

```{r}
dyn.load("/opt/ohpc/pub/apps/libpng/1.6.37/lib/libpng16.so.16")
```
```{r message=FALSE}
library(seriation)
library(dplyr)
library(purrr)
library(ggplot2)
library(phyloseq)
library(microViz)
library(shiny)
```

```{r}
mice <- readRDS('/xdisk/bhurwitz/bh_class/data/microviz/mice.rds')
shao19 <- readRDS('/xdisk/bhurwitz/bh_class/data/microviz/shao19.rds')
mice <- tax_fix(mice, verbose = FALSE)
shao19 <- tax_fix(shao19, verbose = FALSE)
```

# Beta diversity (part 1)

So far, this has been relatively straightforward, at least conceptually.
But there is a lot more interesting stuff we can do with microbiome data.

We've looked at one sample at a time and calculated and compared simple summary measures of sample alpha-diversity.

Alpha diversity is sometimes referred to as "within sample" diversity.

Now we're going to look at "beta diversity", or "between sample" diversity.

For this part we're going to move to another dataset.
So you get a little bit more practice examining a phyloseq object.
Look at the rank names, sample data variables etc.

```{r}
shao19 # this object has another part!
```

## Filtering

First, we need to talk about filtering.

### Sample filtering

You should check if any of your samples have a surprisingly low total number of (classified) reads.
This can suggest that something went wrong in the lab (or during sample collection) and the data from this sample might be unreliable.

You might already do this check for total reads and remove poor quality samples during the fastq file processing.

```{r, fig.height=3, fig.width=6}
shao19 %>% 
  ps_mutate(reads = sample_sums(shao19)) %>% 
  samdat_tbl() %>% 
  ggplot(aes(x = reads)) + 
  geom_freqpoly(bins = 500) +
  geom_rug(alpha = 0.5) +
  scale_x_log10(labels = scales::label_number()) +
  labs(x = 'Number of classified reads', y = NULL) +
  theme_bw()
```

How many is enough?
There is no easy answer.

These samples have great depth.
There are a few with much less reads than the rest, and a few with under a million.
You might consider dropping the samples with under a million reads, to see if it affects your results, but in this case we won't.

But 100,000 is still a lot, compared to what older sequencing machines produced: 1000 reads might have been considered very good.
So look at the distribution for your data, in case there are obvious outliers, and look at recent papers using a similar sequencing technique for what kind of threshold they used.

There might also be relevant information for the type of sequencer you used on e.g.
Illumina website.
e.g. for this type of sequencing Illumina suggests you should expect at least a million reads (and this is good for RNA seq analyses).
<https://support.illumina.com/bulletins/2017/04/considerations-for-rna-seq-read-length-and-coverage-.html>

If you are interested, go back and recreate this plot with the 16S sequencing dataset `mice`.

```{r, fig.height=3, fig.width=6}
# mice %>%
#   ps_mutate(reads = sample_sums(mice)) %>%
#   samdat_tbl() %>%
#   ggplot(aes(x = reads)) +
#   geom_freqpoly(bins = 30) +
#   ... 
```

### Taxon filtering

Okay, so we might remove "bad" samples, but how can a taxon be "bad"?

We probably want to filter out **rare** taxa, before performing some kinds of analysis.

#### Why remove rare taxa?

**Rare taxa might be:**

1.  Sequencing errors
2.  Statistically problematic
3.  Biologically irrelevant

**More elaboration for presentation/discussion**

1.  Rare taxa might be erroneous: sequencing errors, chimeric sequences...

2.  Statistical challenges:

    -   There are often A LOT of different rare taxa
        -   Sparsity, lots of zeros
        -   Low precision in very low abundance taxa (integer counts)
        -   Multiple comparisons (we'll get to that later)
    -   Having many many taxa in your dataset can really slow down some analyses!

3.  Biological (ir)relevance:

    -   For many (most?) microbiome research questions, rare taxa are less likely to be relevant
    -   For common host health problems and the human/animal microbiome, if a taxon only occurs in every 100th person, it is less likely to be relevant to risk common disease outcomes (?)

I'm sure one can think of many cases where rare taxa are biologically relevant, but problems 1 and 2 are still there.

#### How to remove rare taxa?

**What is rare?** Two main concepts.

-   Low **prevalence** - taxon only detected in a small number of samples in your dataset.
-   Low **abundance** - relatively few reads assigned to that taxon (on average or in total)

Consider the impact of issues 1, 2, and 3.
Let's say we are not interested in unique taxa that occur in fewer than 2% of samples, and they have to have at least 10,000 reads in total across all samples.

```{r}
# before filtering
ntaxa(shao19)

# after filtering
shao19 %>% 
  tax_filter(min_prevalence = 2 / 100, min_total_abundance = 10000) %>% 
  ntaxa()
```

Wow so that would remove **most** of our unique taxa!
What is going on?
Let's make some plots!

```{r}
# make table of summary statistics for the unique taxa in shao19
shaoTaxaStats <- tibble(
  taxon = taxa_names(shao19),
  prevalence = microbiome::prevalence(shao19),
  total_abundance = taxa_sums(shao19)
)
```

```{r}
p <- shaoTaxaStats %>%
  ggplot(aes(total_abundance, prevalence)) +
  geom_point(alpha = 0.5) +
  geom_rug(alpha = 0.1) +
  scale_x_continuous(
    labels = scales::label_number(), name = "Total Abundance"
  ) +
  scale_y_continuous(
    labels = scales::label_percent(), breaks = scales::breaks_pretty(n = 9),
    name = "Prevalence (%)",
    sec.axis = sec_axis(
      trans = ~ . * nsamples(shao19), breaks = scales::breaks_pretty(n = 9),
      name = "Prevalence (N samples)"
    )
  ) +
  theme_bw()
p
```

So most taxa have a low prevalence, and handful have way more reads than most.

Let's label those points to check which taxa are the big time players.

```{r}
p + ggrepel::geom_text_repel(
  data = function(df) filter(df, total_abundance > 1e9 | prevalence > 0.6),
  mapping = aes(label = taxon), size = 2.5, min.segment.length = 0, force = 15
)
```

Those taxa make sense for this dataset of mostly infant gut microbiome samples.

Now let's zoom in on the less abundant taxa by log-transforming the axes.
We'll also add lines indicating the thresholds of 2% prevalence and 10000 reads abundance.

```{r}
shaoTaxaStats %>%
  ggplot(aes(x = total_abundance, y = prevalence)) +
  geom_vline(xintercept = 10000, color = "red", linetype = "dotted") +
  geom_hline(yintercept = 2 / 100, color = "red", linetype = "dotted") +
  geom_point(alpha = 0.5) +
  geom_rug(alpha = 0.1) +
  scale_x_log10(labels = scales::label_number(), name = "Total Abundance") +
  scale_y_log10(
    labels = scales::label_percent(), breaks = scales::breaks_log(n = 9),
    name = "Prevalence (%)",
    sec.axis = sec_axis(
      trans = ~ . * nsamples(shao19), breaks = scales::breaks_log(n = 9),
      name = "Prevalence (N samples)"
    )
  ) +
  theme_bw()
```

We can break this down by phylum if we add the taxonomic table information.

```{r, fig.height = 5, fig.width=8}
# don't worry about this code if it's confusing, just focus on the plot output
shao19 %>%
  tax_table() %>%
  as.data.frame() %>%
  as_tibble(rownames = "taxon") %>%
  left_join(shaoTaxaStats, by = "taxon") %>%
  add_count(phylum, name = "phylum_count", sort = TRUE) %>%
  mutate(phylum = factor(phylum, levels = unique(phylum))) %>% # to fix facet order
  mutate(phylum = forcats::fct_lump_n(phylum, n = 5)) %>%
  mutate(phylum = forcats::fct_explicit_na(phylum, na_level = "Other")) %>%
  ggplot(aes(total_abundance, prevalence)) +
  geom_vline(xintercept = 10000, color = "red", linetype = "dotted") +
  geom_hline(yintercept = 2 / 100, color = "red", linetype = "dotted") +
  geom_point(alpha = 0.5, size = 1) +
  geom_rug(alpha = 0.2) +
  scale_x_log10(
    labels = scales::label_log(), breaks = scales::breaks_log(n = 5),
    name = "Total Abundance"
  ) +
  scale_y_log10(
    labels = scales::label_percent(), breaks = scales::breaks_log(n = 9),
    name = "Prevalence (%)",
    sec.axis = sec_axis(
      trans = ~ . * nsamples(shao19), breaks = scales::breaks_log(n = 9),
      name = "Prevalence (N samples)"
    )
  ) +
  facet_wrap("phylum") +
  theme_bw(10)
```

**How to pick a threshold?**

Depends on what analysis method you are filtering for!

-   alpha diversity = DO NOT FILTER
-   beta diversity = relevance of threshold depends on your distance measure (next topic!)
-   differential abundance testing = stringent filtering, prevalence \>5%, \>10%? (last topic!)

# Session info

Records your package versions etc.
Useful for debugging / reproducing analysis.

```{r}
devtools::session_info()
```
