<!DOCTYPE html>
<html>
<head>
<title>ex16_tackling_diversity.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h2 id="plunge-into-diversity">Plunge into diversity</h2>
<p>Species diversity, in its simplest definition, is the number of species in a particular area and their relative abundance (evenness).
Once we know the taxonomic composition of our metagenomes, we can do diversity analyses.
Here we will discuss the two most used diversity metrics, α diversity (within one metagenome) and β (across metagenomes).</p>
<ul>
<li>α Diversity: Can be represented only as richness (<em>, i.e.,</em> the number of different species in an environment), or it can be measured considering the
abundance of the species in the environment as well (<em>i.e.,</em> the number of individuals of each species inside the environment). To measure α-diversity, we use indexes such as Shannon's, Simpson's, Chao1, etc.</li>
</ul>
<a href="../fig/03-08-01.png">
  <img src="../fig/03-08-01.png" alt="Alpha diversity diagram: In lake A, we have three fishes, each one of a different species. On lake B, we have two fish, each of a different species. Moreover, we have four fish in lake C, each of different species." />
</a>
<em> Figure 1. Alpha diversity is calculated according to fish diversity in a pond. Here, alpha diversity is represented in its simplest way: Richness. <em/>
<ul>
<li>β diversity is the difference (measured as distance) between two or more environments.
It can be measured with metrics like Bray-Curtis dissimilarity, Jaccard distance, or UniFrac distance, to name a few. Each one
of this measures are focused on a characteristic of the community (<em>e.g.,</em> Unifrac distance measures the phylogenetic relationship
between the species of the community).</li>
</ul>
<p>In the next example, we will look at the α and the β components of the diversity of a
dataset of fishes in three lakes. The most simple way to calculate the β-diversity
is to calculate the distinct species between two lakes (sites). Let us take as an example the diversity between
Lake A and Lake B. The number of species in Lake A is 3. To this quantity, we will subtract the number of these species that are shared with the Lake
B: 2. So the number of unique species in Lake A compared to Lake B is (3-2) = 1. To
this number, we will sum the result of the same operations but now take Lake B as
our reference site. In the end, the β diversity between Lake A and Lake B is
(3-2) + (3-2) = 2. This process can be repeated, taking each pair of lakes as the
focused sites.</p>
<a href="../fig/03-08-02.png">
  <img src="../fig/03-08-02.png" alt=" Alpha and Beta diversity diagram: Each Lake has a different number of species, and each species has a different number of fish individuals. Both metrics are taken into account to measure alfa and beta diversity." />
</a>
<em> Figure 2. Alpha and beta diversity indexes of fishes in a pond.<em/>
<p>If you want to read more about diversity, we recommend to you this <a href="https://link.springer.com/article/10.1007/s00442-010-1812-0">paper</a> on
the concept of diversity.</p>
<h2 id="%CE%B1-diversity">α diversity</h2>
<p>|-------------------+-----------------------------------------------------------------------------------------------------------------|<br>
| Diversity Indices |                             Description                                                                         |<br>
|-------------------+-----------------------------------------------------------------------------------------------------------------|<br>
|      Shannon (H)  | Estimation of species richness and species evenness. More weight on richness.                                   |<br>
|-------------------+-----------------------------------------------------------------------------------------------------------------|<br>
|    Simpson's (D)  |Estimation of species richness and species evenness. More weigth on evenness.                                    |<br>
|-------------------+-----------------------------------------------------------------------------------------------------------------|<br>
|     Chao1         | Abundance based on species represented by a single individual (singletons) and two individuals (doubletons).    |<br>
|-------------------+-----------------------------------------------------------------------------------------------------------------|</p>
<ul>
<li>Shannon (H):</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align:center">Variable</th>
<th style="text-align:center">Definition</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">$ H = - \sum_{i=1}^{S} p_{i} \ln{p_{i}} $</td>
<td style="text-align:center">Definition</td>
</tr>
<tr>
<td style="text-align:center">$ S $</td>
<td style="text-align:center">Number of OTUs</td>
</tr>
<tr>
<td style="text-align:center">$ p_{i} $</td>
<td style="text-align:center">The proportion of the community represented by OTU i</td>
</tr>
</tbody>
</table>
<!-- <img src="https://render.githubusercontent.com/render/math?math=H=-\sum_{i=1}^{S}p_i\:ln{p_i}"> | Definition
<img src="https://render.githubusercontent.com/render/math?math=S"> | Number of OTUs  
<img src="https://render.githubusercontent.com/render/math?math=p_i">|  The proportion of the community represented by OTU i    -->
<ul>
<li>Simpson's (D)</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align:center">Variable</th>
<th style="text-align:center">Definition</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">$ D = \frac{1}{\sum_{i=1}^{S} p_{i}^{2}} $</td>
<td style="text-align:center">Definition</td>
</tr>
<tr>
<td style="text-align:center">$ S $</td>
<td style="text-align:center">Total number of the species in the community</td>
</tr>
<tr>
<td style="text-align:center">$ p_{i} $</td>
<td style="text-align:center">Proportion of community represented by OTU i</td>
</tr>
</tbody>
</table>
<!-- <img src="https://render.githubusercontent.com/render/math?math=D=\frac{1}{\sum_{i=1}^{S}p_i^2}">| Definition   
<img src="https://render.githubusercontent.com/render/math?math=S"> | Total number of the species in the community   
<img src="https://render.githubusercontent.com/render/math?math=p_i" align="middle"> | Proportion of community represented by OTU i     -->
<ul>
<li>Chao1</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align:center">Variable</th>
<th style="text-align:center">Definition</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">$ S_{chao1} = S_{Obs} + \frac{F_{1} \times (F_{1} - 1)}{2 \times (F_{2} + 1)} $</td>
<td style="text-align:center">Count of singletons and doubletons respectively</td>
</tr>
<tr>
<td style="text-align:center">$ F_{1}, F_{2} $</td>
<td style="text-align:center">Count of singletons and doubletons respectively</td>
</tr>
<tr>
<td style="text-align:center">$ S_{chao1}=S_{Obs} $</td>
<td style="text-align:center">The number of observed species</td>
</tr>
</tbody>
</table>
<!-- <img src="../fig/equation.svg">| Definition  
<img src="https://render.githubusercontent.com/render/math?math=F_1,F_2">|Count of singletons and doubletons respectively    
<img src="https://render.githubusercontent.com/render/math?math=S_{chao1}=S_{Obs}">| The number of observed species   -->
 <!-- comment we use https://viereck.ch/latex-to-svg/ to convert from latex to SVG because Chao equation did not render correctly with GitHub math!-->
<h3 id="%CE%B2-diversity">β diversity</h3>
<p>Diversity β measures how different two or more communities are, either in their composition (richness) or in the abundance of the organisms that compose it (abundance).</p>
<ul>
<li>Bray-Curtis dissimilarity: The difference in richness and abundance across environments (samples). Weight on abundance. Measures the differences
from 0 (equal communities) to 1 (different communities)</li>
<li>Jaccard distance: Based on the presence/absence of species (diversity).
It goes from 0 (same species in the community) to 1 (no species in common)</li>
<li>UniFrac: Measures the phylogenetic distance; how alike the trees in each community are.
There are two types, without weights (diversity) and with weights (diversity and abundance)</li>
</ul>
<p>There are different ways to plot and show the results of such analysis. Among others, PCA, PCoA, or NMDS analysis are widely used.</p>
<blockquote>
<h2 id="exercise-1-simple-measure-of-alpha-and-beta-diversities">Exercise 1: Simple measure of alpha and beta diversities.</h2>
<p>In the next picture, there are two lakes with different fish species:
<a href="../fig/03-08-03.png">
<img src="../fig/03-08-03.png" alt="In lake A, we have four different species, two of these species have three specimens each one. This Lake also has two specimens of a third species and only one specimen of a fourth specie. We got nine fish in total. Lake B has only three different species, the most populated species is also present in lake A and has five specimens, and we have only one specimen of each of the other two species. We got seven species total in lake B " />
</a>
<em> Figure 3. <em/></p>
<p>Which of the options below is true for the alpha diversity in lakes A, B, and beta diversity between lakes A and B, respectively?</p>
<ol>
<li>4, 3, 1</li>
<li>4, 3, 5</li>
<li>9, 7, 16</li>
</ol>
<p>Please, paste your result on the collaborative document provided by instructors.
<em>Hic Sunt Leones!</em> (<em>Here be Lions!</em>)</p>
</blockquote>
<details>
  <summary markdown="span">Solution</summary>
  <ul> 
Answer: 2. 4, 3, 5
**Alpha diversity** in this case, is the sum of different species. Lake **A** has **4** different species and lake **B** has **3** different species.
**Beta diversity** refers to the difference between lake A and lake B. If we use the formula in *Figure 2* we can see that to calculate beta diversity, we have to detect the number of species and the number of shared species in both lakes. There is only one shared species, so we have to subtract the number of shared species to the total species and sum the result. In this case, in lake A, we have 4 different species and one shared species with lake B (4-1)=3, and in lake B we have three species and one shared species with lake A (3-1)=2. If we add 3+2, the result is 5. 
</details>
<blockquote>
<h2 id="discussion-indexes-of-diversity">Discussion: Indexes of diversity</h2>
<p>Why do you think we need different indexes to asses diversity? What index will you use to assess the impact of rare, low-abundance taxa?</p>
</blockquote>
<details>
  <summary markdown="span">Solution</summary>
  <ul> 
<p>It will be difficult (if not impossible) to take two communities and observe the same distribution of all members. This outcome is because
there are a lot of <strong>factors</strong> affecting these lineages. Some of the <strong>environmental factors</strong> are temperature, pH, and nutrient concentration.
Also, the interactions of these populations, such as competence, inhibition of other populations, and growth speed,
are an important driver of variation (<strong>biotic factor</strong>). A combination of the <strong>factors mentioned above</strong>, can interact to maintain some populations with low abundance <strong>rare taxa</strong> In order to have
ways to assess hypotheses regarding which of these processes can be affecting the community, we use all these different indexes. Some emphasize the number of species and other the evenness of the OTUs.</p>
<p>To assess the impact of low abundance lineages, one alpha diversity index widely used is the Chao1 index.</p>
</details>
<p>{% include links.md %}</p>

</body>
</html>
